cpp/include/cudf/utilities/type_dispatcher.hpp

```diff
@@ -511,18 +511,9 @@ CUDF_HOST_DEVICE __forceinline__ constexpr decltype(auto) type_dispatcher(cudf::
         std::forward<Ts>(args)...);
     default: {
 #ifndef __CUDA_ARCH__
-      CUDF_FAIL("Unsupported type_id.");
+      CUDF_FAIL("Invalid type_id.");
 #else
-      cudf_assert(false && "Unsupported type_id.");
-
-      // The following code will never be reached, but the compiler generates a
-      // warning if there isn't a return value.
-
-      // Need to find out what the return type is in order to have a default
-      // return value and solve the compiler warning for lack of a default
-      // return
-      using return_type = decltype(f.template operator()<int8_t>(std::forward<Ts>(args)...));
-      return return_type();
+      CUDF_UNREACHABLE("Invalid type_id.");
 #endif
```

Comment:
Would it be useful to make this into a single macro (maybe this should be `CUDF_UNREACHABLE`, so it covers both host and device code)? I see the pattern in a few places in the PR.

Reply:
I considered that, but I didn't want to hide the dependence on `__CUDA_ARCH__`. Failure/raising an error and unreachable code mean very different things in my opinion, and I didn't want to conflate them by replacing this with an idiom that has potential for misuse. What do you think?

Reply:
I'm not sure. It's weird because we do have the uneven handling between host and device as it is. Maybe it should be the other way around, and `CUDF_FAIL` can call `CUDF_UNREACHABLE` if in device code. As in - "we failed on the device, here's an assert if debug and don't expect a return".

Reply:
Tagging @jrhemstad for thoughts on this. I would defer that change to a later PR if possible.

Reply:
I think I'm still in favor of keeping these macros separate. Letting `CUDF_FAIL` defer to an unreachable path seems dangerous. Developers that see `CUDF_FAIL` should be able to reasonably expect an error, and should not use it to signify branches that can be optimized out as impossible to reach. A macro named something like `CUDF_IMPOSSIBLE` might be a compromise, but I think it obscures the intention (in harmful ways) more than it helps with cleanliness/brevity.

Reply:
Yeah, obscuring the intention is the main issue I can see.

Here's what bugs me: we are using `CUDF_UNREACHABLE` both for truly unreachable code and failure. Ideally,  `CUDF_UNREACHABLE` macro would call GCC's `__builtin_unreachable()` if in host code. But we call `CUDF_FAIL` instead in such cases. 
Feels like code that should not be executed should use `CUDF_FAIL` (both host and device) and truly unreachable code should use  `CUDF_UNREACHABLE` (both host and device). I understand that this may do more hard than good, just bringing it up for consideration.

Reply:
I believe all the cases handled in this way are actually unreachable (by enum exhaustion, in most cases). We’re just taking the opportunity to raise an error on the host because we can.